File: src/test/selenium-csharp-nunit/OrangeHrm.SeleniumNUnit/Driver/WebDriverFactory.cs
using System;
using OrangeHrm.SeleniumNUnit.Config;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Firefox;

namespace OrangeHrm.SeleniumNUnit.Driver
{
    /// <summary>
    /// WebDriver factory. This is intentionally conservative:
    /// - In discovery-only mode it returns null and does not touch real browser drivers.
    /// - In real mode it creates a local Chrome/Firefox driver (drivers must already be on PATH).
    /// </summary>
    public static class WebDriverFactory
    {
        // PUBLIC_INTERFACE
        /// <summary>
        /// Creates an IWebDriver if discovery-only is disabled; otherwise returns null.
        ///
        /// In real mode, this expects a driver binary on PATH:
        /// - chromedriver (for Chrome)
        /// - geckodriver (for Firefox)
        /// </summary>
        public static IWebDriver? CreateIfEnabled(OrangeHrmSettings settings)
        {
            if (settings.DiscoveryOnly)
            {
                return null;
            }

            return settings.Browser switch
            {
                "firefox" => CreateFirefox(settings.Headless),
                _ => CreateChrome(settings.Headless)
            };
        }

        private static IWebDriver CreateChrome(bool headless)
        {
            var options = new ChromeOptions();

            if (headless)
            {
                // Use modern headless mode when available
                options.AddArgument("--headless=new");
            }

            options.AddArgument("--window-size=1280,800");
            options.AddArgument("--no-sandbox");
            options.AddArgument("--disable-dev-shm-usage");

            return new ChromeDriver(options);
        }

        private static IWebDriver CreateFirefox(bool headless)
        {
            var options = new FirefoxOptions();
            if (headless)
            {
                options.AddArgument("-headless");
            }

            return new FirefoxDriver(options);
        }
    }
}

File: src/test/selenium-java-junit/src/test/java/org/orangehrm/selenium/driver/WebDriverFactory.java
package org.orangehrm.selenium.driver;

import org.orangehrm.selenium.config.OrangeHrmConfig;
import org.orangehrm.selenium.config.RunMode;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.firefox.FirefoxOptions;
import org.openqa.selenium.remote.RemoteWebDriver;

import java.net.URI;

/**
 * WebDriver factory.
 *
 * Default behavior is discovery-only to prevent any real browser launch during CI validation.
 */
public final class WebDriverFactory {
    private WebDriverFactory() {}

    // PUBLIC_INTERFACE
    public static WebDriver create(OrangeHrmConfig config) {
        /**
         * Create WebDriver.
         *
         * In discovery-only mode, returns a DiscoveryOnlyWebDriver.
         * In real mode, attempts to create a local Chrome/Firefox driver (requires driver binaries on PATH).
         *
         * @param config test configuration
         * @return WebDriver instance
         */
        if (RunMode.isDiscoveryOnly()) {
            return new DiscoveryOnlyWebDriver();
        }

        String browser = (config.getBrowser() == null ? "chrome" : config.getBrowser()).toLowerCase();
        boolean headless = config.isHeadless();

        try {
            if ("chrome".equals(browser)) {
                ChromeOptions options = new ChromeOptions();
                if (headless) {
                    // Mirrors Node suite preference: '--headless=new' where available
                    options.addArguments("--headless=new");
                }
                options.addArguments(
                        "--no-sandbox",
                        "--disable-dev-shm-usage",
                        "--disable-gpu",
                        "--window-size=1440,900"
                );
                return new org.openqa.selenium.chrome.ChromeDriver(options);
            }

            if ("firefox".equals(browser)) {
                FirefoxOptions options = new FirefoxOptions();
                if (headless) {
                    options.addArguments("-headless");
                }
                return new org.openqa.selenium.firefox.FirefoxDriver(options);
            }
        } catch (Exception e) {
            throw new IllegalStateException("Failed to create real WebDriver. Ensure drivers are available on PATH.", e);
        }

        throw new IllegalArgumentException("Unsupported BROWSER=\"" + browser + "\". Use \"chrome\" or \"firefox\".");
    }
}
